File: plugins/com.googlecode.contraildb.core/src/com/googlecode/contraildb/core/async/TaskTracker.java
package com.googlecode.contraildb.core.async;

import java.util.ArrayList;

import kilim.Pausable;

/** 
 * A simple utility for running tasks and waiting until they're all done.
 * @author ted.stockwell
 */
@SuppressWarnings("rawtypes")
public class TaskTracker {
	private ArrayList<IResult> _tasks= new ArrayList<IResult>();

	public <T> IResult<T> submit(ContrailTask<T> task) {
		IResult<T> result= task.submit();
		_tasks.add(result);
		return result;
	}
	
	public <T> IResult<T> track(IResult<T> result) {
		_tasks.add(result);
		return result;
	}
	public void track(IResult... results) {
		for (IResult result:results) {
			_tasks.add(result);
		}
	}
	
	/**
	 * Does not return until all tasks are complete.
	 * @throws a runtime exception if an error occurs in any of the tasks.
	 */
	public void get() throws Pausable {
		TaskUtils.combineResults(_tasks).get();
	}
	/**
	 * Blocking version of get
	 */
	public void getb() {
		TaskUtils.combineResults(_tasks).getb();
	}
}

File: plugins/com.googlecode.contraildb.core/src/com/googlecode/contraildb/core/async/TaskTracker.scala
package com.googlecode.contraildb.core.async;

import java.util.ArrayList;
import scala.collection.mutable.ArrayBuffer


/** 
 * A simple utility for running tasks and waiting until they're all done.
 * This task scheduler executes all received tasks in parallel
 * 
 * @author ted.stockwell
 */
class TaskTracker extends TaskScheduler {
	val _tasks= new ArrayBuffer[Result[_ <: Any]]();
	
	def submit[V](taskMethod: => V):Result[V]= {
	  this.synchronized {
		  val task:Task[V]= Task.toTask(taskMethod);
		  submit(task);
		  return task;
	  }
	}
	
	def track[T](result:Result[T]):Result[T]= {
	  this.synchronized {
		_tasks+= result;
		return result;
	  }
	}

	/**
	 * Executes a function when all the tasks currently being tracked have completed.
	 */
	def onDone (todo: => Any)= {
	  this.synchronized {
	    TaskUtils.combineResults(_tasks) | todo;
	  }
	}
	
	def |(handler: => Any) { onDone(handler) }
	
}
